File: results/1.0/ARP/arp-gpt-4o.c
#include <stdio.h>
#include <stdlib.h>
#include <hammer/hammer.h>

HParser *arp_packet_parser() {
    // Define parsers for all fields of the ARP packet
    HParser *htype = h_uint16();   // Hardware Type
    HParser *ptype = h_uint16();   // Protocol Type
    HParser *hlen = h_uint8();     // Hardware Length
    HParser *plen = h_uint8();     // Protocol Length
    HParser *oper = h_uint16();    // Operation
    HParser *sha = h_repeat_n(h_uint8(), 6);  // Sender Hardware Address (Ethernet uses 6 bytes)
    HParser *spa = h_repeat_n(h_uint8(), 4);  // Sender Protocol Address (IPv4 uses 4 bytes)
    HParser *tha = h_repeat_n(h_uint8(), 6);  // Target Hardware Address (Ethernet uses 6 bytes)
    HParser *tpa = h_repeat_n(h_uint8(), 4);  // Target Protocol Address (IPv4 uses 4 bytes)

    // Aggregate all parsers into a sequence that matches the ARP packet structure
    return h_sequence(htype, ptype, hlen, plen, oper, sha, spa, tha, tpa, NULL);
}

void parse_arp_file(const char *filename) {
    FILE *file = fopen(filename, "rb");
    if (!file) {
        fprintf(stderr, "Could not open file %s\n", filename);
        exit(EXIT_FAILURE);
    }

    fseek(file, 0, SEEK_END);
    long fsize = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t *buffer = (uint8_t *)malloc(fsize);
    if (!buffer) {
        fprintf(stderr, "Memory allocation failed\n");
        fclose(file);
        exit(EXIT_FAILURE);
    }

    if (fread(buffer, 1, fsize, file) != fsize) {
        fprintf(stderr, "Error reading file %s\n", filename);
        free(buffer);
        fclose(file);
        exit(EXIT_FAILURE);
    }
    fclose(file);

    HParser *arp_parser = arp_packet_parser();
    HParseResult *result = h_parse(arp_parser, buffer, fsize);

    if (result->bit_length >= 0) {
        printf("ARP packet parsed successfully.\n");
        // Display parsed fields (this can be expanded or refined as needed)
    } else {
        fprintf(stderr, "Failed to parse ARP packet.\n");
    }

    h_parse_result_free(result);
    free(buffer);
    // No need to explicitly free the parser itself with Hammer; parsers are managed automatically
    // If you do need to dynamically free a custom parser, it might require a different approach
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <binary file>\n", argv[0]);
        return EXIT_FAILURE;
    }

    parse_arp_file(argv[1]);

    return EXIT_SUCCESS;
}

File: results/1.0/ARP/arp-gpt-4o.rs
use nom::{
    bytes::complete::take,
    combinator::map,
    sequence::tuple,
    IResult,
};
use std::env;
use std::fs;

#[derive(Debug)]
struct ArpPacket {
    htype: u16,
    ptype: u16,
    hlen: u8,
    plen: u8,
    oper: u16,
    sha: Vec<u8>,
    spa: Vec<u8>,
    tha: Vec<u8>,
    tpa: Vec<u8>,
}

fn parse_u16(input: &[u8]) -> IResult<&[u8], u16> {
    let (input, bytes) = take(2usize)(input)?;
    Ok((input, u16::from_be_bytes([bytes[0], bytes[1]])))
}

fn parse_u8(input: &[u8]) -> IResult<&[u8], u8> {
    let (input, byte) = take(1usize)(input)?;
    Ok((input, byte[0]))
}

fn parse_arp(input: &[u8]) -> IResult<&[u8], ArpPacket> {
    let (input, (htype, ptype, hlen, plen, oper)) =
        tuple((parse_u16, parse_u16, parse_u8, parse_u8, parse_u16))(input)?;
    let (input, sha) = take(hlen)(input)?;
    let (input, spa) = take(plen)(input)?;
    let (input, tha) = take(hlen)(input)?;
    let (input, tpa) = take(plen)(input)?;

    Ok((
        input,
        ArpPacket {
            htype,
            ptype,
            hlen,
            plen,
            oper,
            sha: sha.to_vec(),
            spa: spa.to_vec(),
            tha: tha.to_vec(),
            tpa: tpa.to_vec(),
        },
    ))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <binary file>", args[0]);
        return;
    }

    let filename = &args[1];
    let data = fs::read(filename).expect("Unable to read file");

    match parse_arp(&data) {
        Ok((_, arp_packet)) => {
            println!("{:?}", arp_packet);
        }
        Err(e) => {
            eprintln!("Failed to parse ARP packet: {:?}", e);
        }
    }
}